scripts: twisterlib: Enable multiple simulator support in twister

ithinuel opened the pull request:
This change introduces the ability in twister to select which
emulation/simulation tool to use on the command line.

If none is specified, it will select the first in the list.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
boards: twister: update boards' twister metadata files

Following the update of the schema, this updates all boards with the new
structure.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

golowanow reviewed (dismissed):
my -1 is mostly because of missing `simics` changes done with #77700

ithinuel commented on doc/develop/test/twister.rst:
```diff
@@ -149,7 +149,23 @@ name:
+        - name: custom
+          exec: AnotherBinary
+
+  By default, tests will be executed using the first entry in the simulation array. Another
```
Do you mind if this is kept for a future improvement?

This extra attribute would require runtime sanity checks to verify its unicity & lookup to find the default rather than picking index 0, therefore adding more complexity to this PR.

ithinuel commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -211,23 +213,24 @@ def setup_handler(self, env: TwisterEnv):
 
-            if self.platform.simulation_exec and shutil.which(self.platform.simulation_exec):
-                handler.ready = True
+            handler.ready = True
+            if simulator.exec and not shutil.which(simulator.exec):
+                handler.ready = False
```
oh good catch, thank you! 👍

ithinuel committed:
scripts: twisterlib: Enable multiple simulator support in twister

This change introduces the ability in twister to select which
emulation/simulation tool to use on the command line.

If none is specified, it will select the first in the list.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel commented:
I don't understand why the blackbox tests are failing. Those related to `fvp_baser_aemv8r` will be addressed shortly.

golowanow commented:
> I don't understand why the blackbox tests are failing. 

https://github.com/zephyrproject-rtos/zephyr/actions/runs/11575495586/job/32222231297?pr=79174#step:9:24
```
scripts/tests/twister_blackbox/test_addon.py .F.FFF.......F 
```

It seems because of changes in this PR, for example 
`pytest -vv $ZEPHYR_BASE/scripts/tests/twister_blackbox/test_addon.py -k test_enable_ubsan `
this test expects to fail at runtime because of ubsan check, but for this PR it now pass as build only on `native_sim`

on main (c22233a1afcb97) fail as expected
```
ERROR   - native_sim/native         scripts/tests/twister_blackbox/test_data/tests/san/ubsan/san.ubsan  FAILED : Failed
INFO    - /tmp/pytest-of-rootroot/pytest-877/test_enable_ubsan_ubsan_0/blackbox-out-container/blackbox-out/native_sim_native/scripts/tests/twister_blackbox/test_data/tests/san/ubsan/san.ubsan/handler.log
ERROR   - *** Booting Zephyr OS build v4.0.0-rc1-45-gc22233a1afcb ***
Running TESTSUITE a1_1_tests
===================================================================
START - test_assert
/home/rootroot/zephyrproject/zephyr/scripts/tests/twister_blackbox/test_data/tests/san/ubsan/src/main.c:25:4: runtime error: signed integer overflow: 2147483647 + 256 cannot be represented in type 'int'

INFO     twister:runner.py:1433 1 test scenarios (1 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO     twister:reports.py:586 0 of 1 test configurations passed (0.00%), 0 built (not run), 1 failed, 0 errored, 0 skipped with 0 warnings in 11.70 seconds
INFO     twister:reports.py:609 In total 1 test cases were executed, 0 skipped on 1 out of total 860 platforms (0.12%)
INFO     twister:reports.py:618 1 test configurations executed on platforms, 0 test configurations were only built.
```

vs. for this PR - the black_box test passed as build only has no ubsan errror
```
INFO     twister:runner.py:1433 1 test scenarios (1 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO     twister:reports.py:586 0 of 1 test configurations passed (0.00%), 1 built (not run), 0 failed, 0 errored, 0 skipped with 0 warnings in 12.76 seconds
INFO     twister:reports.py:609 In total 0 test cases were executed, 0 skipped on 1 out of total 860 platforms (0.12%)
INFO     twister:reports.py:618 0 test configurations executed on platforms, 1 test configurations were only built.
```

nashif commented on scripts/pylib/twister/twisterlib/platform.py:
```diff
@@ -2,16 +2,44 @@
+        self.exec = data.get("exec")
+
+    def is_runnable(self) -> bool:
+        return (self.name == 'qemu') or bool(self.exec) and bool(shutil.which(self.exec))
```
that does not work for native and other simulators, native_sim now is build_only and will never run, that is bad ;-(

ithinuel replied:
Indeed. Thank you for looking into this. I think I found the culprit and will be pushing an update once I'm done running extra checks.

ruuddw commented on doc/develop/test/twister.rst:
```diff
@@ -149,7 +149,23 @@ name:
+
+      simulation:
+        - name: qemu
+        - name: armfvp
```
Any further guidance/examples on how to use multiple simulators for a single board? So far for ARC, we have different "boards" defined for nSIM simulations (under boards/snps/nsim/... and qemu simulations (boards/qemu/arc/...).
Is the idea to get rid of the boards/qemu and move them under the vendor/architecture instead?

ithinuel replied:
Not familiar with those.

I'd say that if you can build the very same binary and run it on different simulator, then they could be brought together under a single board and list the usable simulators.

mmahadevan108 commented:
I have set the milestone for this PR to be 4.1 as we are only accepting bug fixes and documentation updates for for 4.0

nashif commented:
@ithinuel  please rebase

pdgendt commented on scripts/pylib/twister/twisterlib/constants.py:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env python3
+# vim: set syntax=python ts=4 :
```
A format check has been added to compliance checks, we should not add editor specific details here.

nashif replied:
we also have .editconfig :)

nashif commented:
@golowanow  @hakehuang @PerMac @LukaszMrugala  please review
